syntax = "proto3";
package mathop;

service MathOps {
    rpc MatrixMultiply (MatrixOperationRequest) returns (MatrixOperationReply);
    rpc ListAverage (ListAverageOperationRequest) returns (ListAverageOperationReply);
    rpc OrderList (OrderListRequest) returns (OrderListReply);
    rpc ListMax (ListMaxRequest) returns (ListMaxReply);
    rpc ListMin (ListMinRequest) returns (ListMinReply);
    // rpc ListMode (ListModeRequest) returns (ListModeReply);
    // rpc ListMedian (ListMedianRequest) returns (ListMedianReply);
}

message MatrixRow {
  repeated double values = 1;
}

message Matrix {
  repeated MatrixRow rows = 1;
}

message MatrixOperationRequest {
  Matrix matrix_a = 1;
  Matrix matrix_b = 2;
  string operation_id=3;
}

message MatrixOperationReply {
  Matrix result_matrix = 1;
  string status_message = 2; // e.g., "Success", "Matrices have incompatible dimensions"
}

message ListAverageOperationRequest{
  repeated double x = 1;
    string operation_id=2;

}

message ListAverageOperationReply{
  optional double result_average = 1;
  string status_message = 2; // e.g., "Success".

}


message OrderListRequest{
  repeated double x = 1;
  bool ascending = 2; // true for ascending, false for descending
  string operation_id=3;

}

message OrderListReply{
  repeated double result = 1;
  string status_message = 2; // e.g., "Success".

}


message ListMaxRequest{
  repeated double x = 1;
    string operation_id=2;

}

message ListMaxReply{
  optional double result = 1;
  string status_message = 2; // e.g., "Success".

}


message ListMinRequest{
  repeated double x = 1;
    string operation_id=2;

}

message ListMinReply{
  optional double result = 1;
  string status_message = 2; // e.g., "Success".

}


// message ListModeRequest{
//   repeated double x = 1;
//     string operation_id=2;

// }

// message ListModeReply{
//   optional double result = 1;
//   string status_message = 2; // e.g., "Success".

// }

// message ListMedianRequest{
//   repeated double x = 1;
//     string operation_id=2;

// }

// message ListMedianReply{
//   optional double result = 1;
//   string status_message = 2; // e.g., "Success".

// }